backbone/src/main/java/org/researchstack/backbone/step/active/recorder/SensorRecorder.java

```diff
@@ -86,54 +87,66 @@ public void start(Context context) {
                 }
                 anySucceeded |= success;
+
+                if (!success) {
+                    LogExt.i(SensorRecorder.class, "Failed to register sensor: " + sensor);
+                }
             }
         }
 
         if (!anySucceeded) {
-            super.onRecorderFailed("Failed to initialize sensor");
+            super.onRecorderFailed("Failed to initialize any sensor");
         } else {
             super.startJsonDataLogging();
         }
-
-        startRunnableForWritingJson();
     }
 
-    protected void startRunnableForWritingJson() {
-        // These will be used to monitor periodic writes to get an accurate write frequency
-        mainHandler = new Handler();
-        writeCounter = 1;
-        writeStartTime = System.currentTimeMillis();
+    @Override
+    public final void onSensorChanged(SensorEvent sensorEvent) {
 
-        if (frequency < 0) {
-            // Writing the JSON will be done manually, and not at a specific frequency
-        } else {
-            writeDelayGoal = calculateDelayBetweenSamplesInMilliSeconds();
-            jsonWriterRunnable = new Runnable() {
-                @Override
-                public void run() {
-                    writeJsonData();
-
-                    writeCounter++;
-                    // Offset delay from the writeJsonData call to get an accurate write frequency
-                    long delayGoal = ((writeStartTime + (writeCounter * writeDelayGoal)) - System.currentTimeMillis());
-                    // The device is not fast enough to keep up, so we will get a frequency only
-                    // as fast as it can do, so just make the delay goal be the original delay
-                    if (delayGoal <= 0) {
-                        // minimal write delay to give the UI thread some time to catch up
-                        // and hopefully get the frequency back up to the desired one
-                        delayGoal = 1;
-                    }
-
-                    mainHandler.postDelayed(jsonWriterRunnable, delayGoal);
-                }
-            };
-            mainHandler.postDelayed(jsonWriterRunnable, writeDelayGoal);
+        JsonObject jsonObject = new JsonObject();
+
+        if (timestampZeroReferenceNanos <= 0) {
+            // set timestamp reference, which timestamps are measured relative to
+            timestampZeroReferenceNanos = sensorEvent.timestamp;
+
+            // record date equivalent of timestamp reference
+            long uptimeNanos;
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
+                uptimeNanos = SystemClock.elapsedRealtimeNanos();
+            } else {
+                uptimeNanos = SystemClock.elapsedRealtime() * 100000; // millis to nanos
+            }
+
+            long timestampReferenceMillis = System.currentTimeMillis()
+                    + ((timestampZeroReferenceNanos - uptimeNanos) / 1000000L);
```

Comment:
This value 1000000L has 6 zeros, while the value above has 5 zeros.  Intentional?

Reply:
Thanks